demos/fleet_provisioning/fleet_provisioning_with_csr/fleet_provisioning_csr_demo.c

```diff
@@ -0,0 +1,1271 @@
+#endif
+#ifndef CLAIM_PRIVATE_KEY_PATH
+    #error "Please define path to claim private key (CLAIM_PRIVATE_KEY_PATH) in demo_config.h."
+#endif
+#ifndef PROVISIONING_PRIVATE_KEY_PATH
+    #error "Please define path to private key to provision (PROVISIONING_PRIVATE_KEY_PATH) in demo_config.h."
+#endif
+#ifndef PROVISIONING_CSR_PATH
+    #error "Please define path to CSR to use to provision (PROVISIONING_CSR_PATH) in demo_config.h."
+#endif
+#ifndef PROVISIONING_CERT_PATH
+    #error "Please define path to which to write the obtained certificate (PROVISIONING_CERT_PATH) in demo_config.h."
+#endif
+
+/**
+ * @brief The length of #PROVISIONING_TEMPLATE_NAME.
+ */
+#define PROVISIONING_TEMPLATE_NAME_LENGTH    ( ( uint16_t ) ( sizeof( PROVISIONING_TEMPLATE_NAME ) - 1 ) )
+
+/**
+ * @brief Size of AWS IoT Thing name buffer.
+ *
+ * See https://docs.aws.amazon.com/iot/latest/apireference/API_CreateThing.html#iot-CreateThing-request-thingName
+ */
+#define MAX_THING_NAME_LENGTH                128
+
+/**
+ * @brief Number of seconds to wait for response from AWS IoT Fleet
+ * Provisioning APIs.
+ */
+#define FLEET_PROV_RESPONSE_WAIT_SECONDS     ( 2 )
+
+/**
+ * @brief The maximum number of times to run the loop in this demo.
+ *
+ * @note The demo loop is attempted to re-run only if it fails in an iteration.
+ * Once the demo loop succeeds in an iteration, the demo exits successfully.
+ */
+#ifndef FLEET_PROV_MAX_DEMO_LOOP_COUNT
+    #define FLEET_PROV_MAX_DEMO_LOOP_COUNT    ( 3 )
+#endif
+
+/**
+ * @brief Time in seconds to wait between retries of the demo loop if
+ * demo loop fails.
+ */
+#define DELAY_BETWEEN_DEMO_RETRY_ITERATIONS_SECONDS    ( 5 )
+
+/**
+ * @brief Size of buffer in which to hold the certificate.
+ */
+#define CERT_BUFFER_LENGTH                             2048
+
+/**
+ * @brief Size of buffer in which to hold the certificate id.
+ *
+ * See https://docs.aws.amazon.com/iot/latest/apireference/API_Certificate.html#iot-Type-Certificate-certificateId
+ */
+#define CERT_ID_BUFFER_LENGTH                          64
+
+/**
+ * @brief Size of buffer in which to hold the certificate ownership token.
+ */
+#define OWNERSHIP_TOKEN_BUFFER_LENGTH                  512
+
+/**
+ * @brief Status values of the Fleet Provisioning response.
+ */
+typedef enum
+{
+    ResponseNotReceived,
+    ResponseAccepted,
+    ResponseRejected
+} ResponseStatus_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Status reported from the MQTT publish callback.
+ */
+static ResponseStatus_t responseStatus;
+
+/**
+ * @brief Buffer to hold the provisioned AWS IoT Thing name.
+ */
+static char thingName[ MAX_THING_NAME_LENGTH ];
+
+/**
+ * @brief Length of the AWS IoT Thing name.
+ */
+static size_t thingNameLength;
+
+/**
+ * @brief Buffer to hold responses received from the AWS IoT Fleet Provisioning APIs.
+ */
+static char payloadBuffer[ NETWORK_BUFFER_SIZE ];
+
+/**
+ * @brief Length of the payload stored in #payloadBuffer.
```

Comment:
Why is a global for the payload size required? If required. it will be helpful to have the reasoning in the comment